### DISCUSSION TITLE
Proposal: "Wrapper" component for groups of routes in folders

### DISCUSSION THREAD

**Discussion by cassidoo** [vercel/next.js#26389] [Help] [answered]
### Describe the feature you'd like to request

Right now, it's difficult to share state across only *some* routes. There's a very "all or nothing" approach to sharing state with just React's `useState`/`useReducer` + Context in `_app.js`. You can work around that (or use libraries like swr and react-query), but it's not very intuitive nor ergonomic.

I'm proposing that there's some kind of "wrapper" component for groups of routes in folders, as seen below.

### Describe the solution you'd like

I'm proposing a folder structure similar to this example:

```
pages/
  examplefolder/
    _index.js
    route1.js
    route2.js
  examplefolder2/
    examplefolder3/
      _index.js
      route3a.js
      route3b.js
    _index.js
    route4.js
    route5.js
  _app.js
  index.js
  ...

```

If this could exist in Next.js, where `_index.js` (or some other file name) could "wrap" the other files in that `example` folder, that would be a really sleek solution to sharing state across groups of routes.

As a developer, I would expect the `_index.js` files would act on files in their folder just like `_app.js` does for an entire Next.js application. A developer would able to wrap that component in a Context, or query data, or set up a layout from there.

A solution like this would make for almost a "microsite" type of structure across folders in an app, allowing developers to intuitively make shared layouts and state and have a solid separation of application "sections".

### Describe alternatives you've considered

There are some escape hatches built in now for this that aren't very intuitive. The easiest solutions are using swr and react-query, but it would be nice to have a more "vanilla React" way of doing it.

There is something similar to this in [Nuxt.js](https://nuxtjs.org/examples/routing-nested-pages) for a potential API alternative, as well!

**Comment by antoinegag** [comment]
Would be cool if that wrapper could also allow for rendering stuff for all routes. 

For exemple rendering a specific header above every route in a folder without having to include it in every single route or a global navbar (and so on).

**Comment by cassidoo** [comment]
> Would be cool if that wrapper could also allow for rendering stuff for all routes.
> 
> For exemple rendering a specific header above every route in a folder without having to include it in every single route or a global navbar (and so on).

Yes @antoinegag! That's what I meant by "set up a layout from there" exactly. Being able to do it this way would reduce the need to use external libraries [like this one](https://github.com/moxystudio/next-layout) to make easy layout groups.

**Comment by gavinsharp** [comment]
I really like the idea of having a built-in mechanism like this one, it's something I've had to re-create in a couple of different Next apps so far.

I wonder whether tying this to directory tree/routing might be overly restrictive, though. I could see people wanting to specify shared wrapper/layouts with more granularity than what this would allow (i.e. on a page-by-page basis). Maybe a named export on the page could be used to reference a wrapper/layout instead? I guess that would make it more annoying to have the same wrapper used on _all_ pages, so maybe that is just a different proposal with a different set of use cases :)

**Comment by apkoponen** [comment]
Would the wrappers in nested _index.js files override or ”merge” with their top-level wrappers?

E.g. if the `examplefolder2/_index.js` has `<ExampleWrapper2 />` and `examplefolder3/_index.js` has `<ExampleWrapper3 />` should `route3a.js` have both or just `<ExampleWrapper3 />`?

**Reply by amaster507** [reply]
My logic would be that it would flow up the tree. So if I had an `_index` or whatever it gets named in a child directory that the page is wrapped in that and then sent along the tree upward all the way to `_app` at the root. This would allow for a native tree structuring and rendering of the app. E.G.: everything under this directory gets rendered with...

**Comment by Skn0tt** [comment]
This is an interesting proposal! My initial reaction was "ugghh, that sounds complicated". I've yet to fully grasp the difference between `_app` and `_document`, and I'd prefer to not have to learn about `_index`.

Though on second though, I'm growing to like it! Especially the Layout part of it, which allows for much easier nested layouts compared to the [getLayout](https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/) pattern.

@apkoponen To enable both nested layouts and deep-nested paths that don't share layout, I'd imagine this to be configurable. Maybe even by importing `ExampleWrapper2` in `ExampleWraper3`:

```ts
import ExampleWrapper2 from "../_index"

export default function ExampleWrapper3(props) {
  return (
    <ExampleWrapper2>
      <h1>This is a header</>
      {props.children}
    </ExampleWrapper2>
}
```

Some more questions:
- would `_index` have support for `getStaticProps` or `getServerSideProps`?
- what are the tradeoffs of this proposal compared to the [getLayout](https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/) pattern?

**Comment by JReinhold** [comment]
You all might want to look into SvelteKit's nested layouts for inspiration which sounds like it is exactly this.

https://kit.svelte.dev/docs#layouts-nested-layouts

EDIT: As mentioned [on Twitter](https://twitter.com/iamandrewluca/status/1406664811555995659?s=19) there's also a Nested Pages feature in Nuxt that can inspire. 
https://nuxtjs.org/examples/routing-nested-pages

**Comment by nandorojo** [comment]
I have a `/dashboard` path with 8 nested screens inside of it: `/dashboard/calendar`, `/dashboard/account`, etc.

These pages all share a common layout that isn’t used anywhere else. It would be awesome to have an `_app.js` used only for those routes. 

At first, to solve this, I conditionally rendered the `/dashboard` layout  component based on `router.pathname` in `_app.tsx`. But it felt wrong for this to be positioned globally, so now I manually wrap each page inside of `/dashboard` with the layout provider. 

This proposal would be a much more elegant solution.

I’d prefer a name other than `_index.js`, since I often use this file name in other parts of my app. But that’s a minor point, and @cassidoo doesn’t seem too tied to the name.

**Reply by gramsco** [reply]
I have the exact same use case and I use the following: ```pages/dashboard/[tab]```

Where tab is whatever ```getStaticPaths``` reads from an array of tabs I give it. That system works fine and you can just "plug" the given tab component and I'm good that way. But it does "feel" hacky (even though it's pretty clean !). 

This proposal goes *further* and I'd say it shines more with state management, scripts, etc than "just" layout. 

Really interesting!

**Comment by PhilipJohnBasile** [comment]
A more “vanilla React" approach would be awesome.

**Comment by kylemh** [comment]
I definitely think we need a more first-class solution to nested layout.

I think tying it to the folder structure would be limiting or at least be a paradigm where it forces devs to structure the paths in their app in a way that allows the sharing of state. For example, if someapp.com/home and someapp.com/data/1 share the same layout or state, I’m not sure how the proposed solution would resolve the issue.

An alternative API might be similar to the “Bonus” solution outlined in [Adam Wathan’s Persistant Layouts](https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/) article.

I could see 2 functions:

1. getLayout
2. getLayoutProvider

The combination of these functions can be used to provide shared UI and shared state in any route; however, at that point maybe you’d only need getLayout since you could put providers there anyway.

**Comment by cherscarlett** [comment]
This would be extremely helpful in building or using component libraries with next.js 🚀

**Comment by vikasg603** [comment]
Very interested and needy solution.
Especially the layout part.

Having constant layout patterns across a folder is very much needed for my application too.

**Comment by cdebotton** [comment]
I would really like to see this in next! We currently attach a `getLayout` method to page components to compose layouts in the `_app` component but this would really cut down on a lot of the repetition.

It would be great to have the ability to control whether nested layouts overrode or composed their parent layouts.

**Comment by amaster507** [comment]
It was amazing literally 5 minutes into converting a react app to next I had this epiphany and thought surely this is supported. I imaged it worked just like `_app` did in the root of pages. I created a new `_app` file in my sub directly and coded up a quick test to confirm it worked, and sadly learned that I was highly mistaken. Now off to beat my head into my desk as I try to implement some form of the getLayout algorithm into our application everywhere.

**Comment by danieljpgo** [comment]
This solution would be very useful for building something like this:

![image](https://user-images.githubusercontent.com/37938090/124535564-b4c8a200-ddec-11eb-8f80-65d046a5ee06.png)

**Reply by danieljpgo** [reply]
But what we really need is a first-class solution to nested layout.

**Reply by vicentereig** [reply]
Supporting nested routes would be phenomenal. It brings me very fond memories of Ember's Routing API.

Besides improving how layouts are organized or rendering just what it's needed, it could have a positive impact when doing SSG as well. For example, in the context of generating large static sites (tens of thousands of pages), child routes wouldn't need to recalculate or refetch their parent's `getStaticProps`. Shaving off some precious build time.

**Comment by leerob** [comment]
**Related:** I just published documentation on how you can [currently handle layouts with Next.js](https://nextjs.org/docs/basic-features/layouts). To be clear, this is documenting what already exists, not solving this proposal.

Thank you for this proposal! We are continuing to investigate built-in layouts support 😄

**Comment by lunelson** [comment]
I love this idea, I've often thought of the `_layout.jade` file-based wrapping that I used years ago in [Harp.js](http://harpjs.com/docs/development/layout), but never thought of it as a way to provide data as well. 

FWIW it looks like Gatsby 4 has caught up to this idea too:

<img width="1629" alt="image" src="https://user-images.githubusercontent.com/1242864/133976775-02f7e72d-28b9-4ec9-894b-6abd281be819.png">

**Comment by tujoworker** [comment]
For **inspiration purposes only** I post my Gatsby Plugin, where you can add layouts – or all kinds of wrappers, even nested wrappers. 

In the README file there are a lot of details.

https://github.com/tujoworker/gatsby-plugin-wrap-pages

I works even inside your own plugin. Which makes it a very interesting to use for [micro frontends](https://github.com/tujoworker/gatsby-plugin-wrap-pages#micro-frontends).

But native `_layout.tsx` support in Gatsby v4 would be amazing.

**Comment by jesstelford** [comment]
The release of [middleware](https://nextjs.org/docs/middleware) shows how this nested-file approach can work quite well!

Personally, I'd love to see nested `_app`s be the mechanism as @amaster507 [mentioned above](https://github.com/vercel/next.js/discussions/26389#discussioncomment-922493) allowing for solving @danieljpgo [example](https://github.com/vercel/next.js/discussions/26389#discussioncomment-968174) like so:

![dashboard](https://user-images.githubusercontent.com/37938090/124535564-b4c8a200-ddec-11eb-8f80-65d046a5ee06.png)

```
.
├── _app.js
├── index.js
└── users
    ├── _app.js
    ├── index.js
    └── [id]
        ├── _app.js
        ├── index.js
        └── edit
            └── index.js
```

The nested `_app`s could also provide a home for nested data fetching via `getStaticProps` etc (but that's a different proposal).

---

*Update*: I've created [`@ceteio/next-layout-loader`](https://github.com/ceteio/next-layout-loader) which is a (kinda clunky) workaround to support nested `_app.js` / `_layout.js` files. Not as automatic as I'd like, but gives a feel for how it might work :+1:

**Reply by lunelson** [reply]
I like this and totally agree that middleware has essentially proven the viability and that they could also be responsible for data fetching, I only think that calling them all `_app.js` might be confusing, I think `_layout.js` is intuitive and has precedent in other frameworks as mentioned above

**Reply by danieljpgo** [reply]
Totally agree, great idea @jesstelford

**Reply by jesstelford** [reply]
I've created [`@ceteio/next-layout-loader`](https://github.com/ceteio/next-layout-loader) which is a (kinda clunky) workaround to support nested `_app.js` / `_layout.js` files. Not as automatic as I'd like, but gives a feel for how it might work :+1:

**Comment by eric-burel** [comment]
`_layout` but for API routes could solve the issue of startup code: https://github.com/vercel/next.js/discussions/11686
In this case the layout would be middlewares common to a set of pages. That would prevent having to call "connectToDb()" in all API routes for instance, instead you'd do it in the API route _layout to feed a context, like we do in graphql. 
The nesting let's you control on which set of route the middleware is applied, eg ignore routes that don't require a connection to the db.

**Comment by mohsen1** [comment]
Passing the path to page component being rendered by `_app.tsx` in the `AppProps` (similar to `__N_SSG`, `__N_SSP` and `__N_RSC`) will probably unlock the issue here. You can decide what layout to render directly in the App. It will look like this:

```tsx
// dynamic-layout.tsx
import React from "react";
import ExampleLayout1 from "../components/ExampleLayout1";
import ExampleLayout2 from "../components/ExampleLayout2";
import DefaultLayout from "../components/DefaultLayout";

const DynamicLayout: React.FC<{ componentPath: string }> = ({
  componentPath,
  children,
}) => {
  if (componentPath.includes("layout1")) {
    return <ExampleLayout1>{children}</ExampleLayout1>;
  }
  if (componentPath.includes("layout2")) {
    return <ExampleLayout2>{children}</ExampleLayout2>;
  }

  return <DefaultLayout>{children}</DefaultLayout>;
};
export default DynamicLayout;
```

```tsx
// _app.tsx
import { AppProps } from "next/app";
import DynamicLayout from "../dynamic-layout";

function MyApp({ Component, pageProps, __N_PAGE_PATH }: AppProps,) {
  return (
    <DynamicLayout componentPath={__N_PAGE_PATH }>
      <Component {...pageProps} />
    </DynamicLayout>
  );
}

export default MyApp;
```

Will this work in your case?

**Reply by jesstelford** [reply]
I had a very similar thought last night, and have been putting together a little lib which uses code generation to achieve what you've written out manually. So it should work with any arbitrary number of nested files & layouts!

I'll get it shipped in the next couple days.

**Comment by jesstelford** [comment]
**I've solved this with [`@ceteio/next-layout-loader`](https://github.com/ceteio/next-layout-loader)!**

It automatically loads `pages/**/_layout.tsx` files, intelligently wrapping them around `pages/_app.tsx` based on the current route:

```javascript
// pages/_app.jsx
const filename = preval`module.exports = __filename`;
const withLayoutLoader = codegen.require("@ceteio/next-layout-loader", filename);

// The magic lives here!
export default withLayoutLoader(({ Component, pageProps }) => (
  <Component {...pageProps} />
));
```

```javascript
// pages/dashboard/_layout.jsx
export default function Layout({ children }) {
  return (
    <div style={{ backgroundColor: "#e2e8f0", padding: "1rem" }}>
      {children}
    </div>
  );
}

// To hide this layout component from the router / build pipeline
export const getStaticProps = async () => ({ notFound: true });
```

```javascript
// pages/dashboard/index.jsx
import NextLink from "next/link";

export default function User() {
  return (
    <div style={{ backgroundColor: "white" }}>
      <NextLink href="/">Home</NextLink>
    </div>
  );
}

export async function getStaticProps() {
  return {
    // props are also passed to _layout components
    props: {
      greeting: "Hello world"
    }
  };
}
```

**Check it out**:
- Repo: https://github.com/ceteio/next-layout-loader
- See it in action on Codesandbox: https://githubbox.com/ceteio/next-layout-loader/tree/main/example

**Comment by msdrigg** [comment]
This is functionality I would very much appreciate

I tried out sveltekit recently, and this one of the bigger niceties of sveltekit that I am missing in nextjs.

**Comment by frei-x** [comment]
What are the alternatives now

**Reply by JClackett** [reply]
Waiting for Next.js to come up with a "unique" name for layout components to include in a "new" release :)

**Comment by jasperfurniss** [comment]
<img src="https://media4.giphy.com/media/MZocLC5dJprPTcrm65/giphy.gif"/>

https://twitter.com/leeerob/status/1521713571377287171?s=20&t=3ANpHD6sbd2n05psZjmbow

**Comment by leerob** [comment]
Hey everyone!

Excited to share we are [actively working on a solution](https://twitter.com/leeerob/status/1521659624516030466) for nested routing and better layouts support. Expect an RFC very soon! 🥳

**Reply by rtrembecky** [reply]
amazing! thanks for the hard work <3

**Comment by cassidoo** [comment]
Hey folks, the RFC has been shared!! 🥳 

#37136

**Reply by JClackett** [reply]
Really really not a fan of the page name convention. Text editor going to be such a mess 😔

**Comment by leerob** [comment, marked as answer]
**Update:** We’re excited to announce [Next.js 13](https://nextjs.org/13), which includes the new `app/` directory (beta) with support for nested layouts, colocation of data fetching, streaming, Server Components by default, and much more.

Inside the `app/` directory, there is a powerful new way to fetch data with React’s `use()` hook and the extended Web `fetch()` API. This allows you to colocate data fetching directly inside components for the most flexibility, including inside layouts.
